Feature/Add LLMs Integration For Oracle Cloud Infrastructure(OCI) Data Science Model Deployment Endpoint

mingkang111 opened the pull request:
- **Description:** 
    - [OCI Data Science](https://docs.oracle.com/en-us/iaas/data-science/using/home.htm) is a fully managed and serverless platform for data science teams to build, train, and manage machine learning models in the Oracle Cloud Infrastructure. This PR add integration for using LangChain with an LLM hosted on a [OCI Data Science Model Deployment](https://docs.oracle.com/en-us/iaas/data-science/using/model-dep-about.htm). To authenticate, [oracle-ads](https://accelerated-data-science.readthedocs.io/en/latest/user_guide/cli/authentication.html) has been used to automatically load credentials for invoking endpoint.
- **Issue:** None
- **Dependencies:** `oracle-ads`
- **Tag maintainer:** @baskaryan
- **Twitter handle:** None

Commit:
added llm integration for oci data science model deployment

Commit:
added import for oci data science model deployment

Commit:
added documentation for oci data science model deployment

Commit:
added authentication section in doc

Commit:
fixed by comments

Commit:
fixed lint warning

Commit:
fixed by comments: fetch endpoint from env var

Commit:
resolved conflict

Commit:
fixed bugs

Commit:
udpated documentation

Commit:
fixed bugs

Commit:
added exclude field to accept more session based fields while validating llm

Commit:
added test for oci model deployment endpoint llm

Commit:
fixed lint

Commit:
fixed tests

Commit:
Merge branch 'master' of github.com:mingkang111/langchain into feature/add_llms_integration_for_oci_datascience_model_deployment_endpoint

mingkang111 commented:
Hi @ baskaryan, I have prepared test evidences page for this PR. It includes test results with endpoint, formate/lint/spell check as informed in contribution guideline. Feel free to let me know if you want to check with it. Thanks for helping with reviewing!

mingkang111 commented on libs/langchain/langchain/llms/oci_data_science_model_deployment_endpoint.py:
```diff
@@ -0,0 +1,369 @@
+
+    def send_request(
+        self,
+        data: Any,
```
![Screenshot 2023-12-05 at 9 58 39 AM](https://github.com/langchain-ai/langchain/assets/49049296/cfaa7902-2a0d-4288-9458-4a35088b0422)
Looks like it cannot recognize Any as type? cc: @baskaryan

hwchase17 replied:
you havent typed **kwargs, that should be `**kwargs: Any`

mingkang111 replied:
Thank you @hwchase17 ! I fixed it.

Commit:
added type hint for kwargs argument

Commit:
fixed lint

Commit:
format

Commit:
docs nits

mingkang111 commented:
Hi @hwchase17 @efriis , thanks for helping with reviewing! May I know is there any change I need to make for this PR? Normally what is the timeline for merging the PR and release feature ( I saw release is ad-hoc from [here](https://github.com/langchain-ai/langchain/blob/master/.github/CONTRIBUTING.md#-release-process), may I know is there any proposed date for the next release) ?  Thanks again for your help!

efriis commented:
Hey @mingkang111 ! Was having some trouble with the OCI signup flow and was unable to create an API key. If that's something you can help with, let me know! Otherwise I'll try again next week. 

We're in the middle of a big refactor that'll make our packages more modular (here's some [context](https://github.com/langchain-ai/langchain/discussions/14243)). because of that we likely won't be landing or releasing any new features until later next week.

mingkang111 commented:
Hi @efriis ! Definitely, I can help with the OCI signup flow. Here is the doc about setting API Keys: [link](https://docs.oracle.com/en-us/iaas/Content/API/Concepts/apisigningkey.htm). Feel free to let me know if there any issues you see when following this guideline. We can also meet through zoom.

Thanks for sharing the information!

streamnsight commented:
@efriis @hwchase17 
Thanks for this great contribution.
I'm with OCI Data and AI team, and I have one concern with the way this is organized, especially the fact that OCI has an official Gen AI Service and is working on contributing the classes for LLM and Embeddings in Langchain.
So you beat us to the punch there, which is appreciated, however you are using the generic model deployment endpoint, not the specific Gen AI endpoints. 
So in terms of the naming conventions, I would refrain from suffixing this work with xxxLLM as these are not LLM endpoints, and that would make it confusing to OCI users.
I think it should just be called BaseModelDeployment, XYZModelDeployment etc... and should support **kwargs so any param could be passed and made generic, and refrain from defining model specific classes. 
If you want to showcase using Llama2 deployed with this endpoint and using it for different use cases, that sounds great, however that should be part of a guide on how to use the ModelDeployment class / endpoints, as it seems like defining specific classes for this purpose and only exposing that in the docs greatly narrows the purpose of the ModelDeployment endpoint and will potentially clash with the official Gen AI service in the future.
Happy to discuss.

Commit:
moved model params out of Base

streamnsight commented:
@efriis @mingkang111 @hwchase17 
I guess I missed the fact that @mingkang111 is also with Oracle, so my apologies for the confusion.
I want to clarify that I am not against this PR, I am just suggesting that if the base class was a ModelDeployment class (and not LLM specific), that would allow other users to deploy their own (generic or specialized) models using that base.
Then specific TGI and VLLM LLM classes based on ModelDeployment sound great for those who want to focus on using those specific models deployments frameworks

Commit:
pop requests kwargs

Commit:
Merge branch 'feature/add_llms_integration_for_oci_datascience_model_deployment_endpoint' of github.com:mingkang111/langchain into feature/add_llms_integration_for_oci_datascience_model_deployment_endpoint

Commit:
fixed merge confict

mingkang111 commented:
Hi @efriis , I just resolved the merge conflicts with the master branch. Can I get your review again? Thanks a lot for your help!

efriis commented:
Thanks! Looks like it's still failing some CI checks if you want to address those.

@streamnsight would appreciate a 👍 or comments from you as well if your team is planning on contributing too! Any structural/naming changes get much more difficult after the initial merge because we maintain backwards compatibility.

Commit:
move to langchain community

Commit:
fixed doc

Commit:
fixed oci model deployment endpoint integration tests

Commit:
removed BaseOCILLM

Commit:
reverted sorting

Commit:
fixed lint

streamnsight commented:
@efriis 
looks good to me. Thanks

Commit:
fixed format

Commit:
fixed lint

Commit:
fixed failed tests

Commit:
fixed pydoc

Commit:
fixed doc preview

Commit:
fixed test for serialization

mingkang111 commented:
> @efriis @mingkang111 @hwchase17 I guess I missed the fact that @mingkang111 is also with Oracle, so my apologies for the confusion. I want to clarify that I am not against this PR, I am just suggesting that if the base class was a ModelDeployment class (and not LLM specific), that would allow other users to deploy their own (generic or specialized) models using that base. Then specific TGI and VLLM LLM classes based on ModelDeployment sound great for those who want to focus on using those specific models deployments frameworks

Hi @streamnsight , thanks for your suggestion. We currently offering OCIModelDeploymentTGI/OCIModelDeploymentVLLM as we provided these two types of deployment (please feel free to explore [here](https://github.com/oracle-samples/oci-data-science-ai-samples/tree/main/model-deployment/containers/llama2#overview)), In our ongoing commitment to enhance user experience, we are excited to integrate these deployments with LangChain, offering our users a seamless and efficient interface. Our dedicated team is in the process of developing more generic deployments, and we will have another PR to accommodate more generic deployments in the future. We are enthusiastic about these developments and believe they will significantly benefit our users.

Commit:
Merge branch 'master' into feature/add_llms_integration_for_oci_datascience_model_deployment_endpoint

mingkang111 commented:
Hi @efriis , thank you for your timely reminder. We are aware of the potential changes on the horizon and wish to assure you that these future updates are designed with great care. Our aim is to ensure they do not impact existing structures or naming conventions, thereby avoiding any backward compatibility issues.  

I have addressed the issues with the failing CI checks and adapted our project to align with the recent changes in LangChain, the transition to `langchain-community`. I would greatly appreciate it if you could review these updates. Your feedback is invaluable to the continuous improvement of our work. 

Thanks again for your time!

Commit:
Merge branch 'master' into feature/add_llms_integration_for_oci_datascience_model_deployment_endpoint

Commit:
Merge branch 'master' into feature/add_llms_integration_for_oci_datascience_model_deployment_endpoint

mingkang111 commented:
Hello @efriis , I noticed there is a release happening soon: https://github.com/langchain-ai/langchain/pull/14867. May I know is that possible to add this change into the release as well? Thank you!

efriis commented:
Hey @mingkang111 ! We actually already missed the community release in #14864 (this will be in community).

I'm actually still running into the same error simply going through the account sign-up flow :(. Could you email me at erick@langchain.dev and we can figure out testing?

![Screenshot 2023-12-18 at 1 29 59 PM](https://github.com/langchain-ai/langchain/assets/9557659/4fe55a75-95aa-4330-8f6f-2ae612bee638)

mingkang111 commented:
> Hey @mingkang111 ! We actually already missed the community release in #14864 (this will be in community).
> 
> I'm actually still running into the same error simply going through the account sign-up flow :(. Could you email me at [erick@langchain.dev](mailto:erick@langchain.dev) and we can figure out testing?
> 
> ![Screenshot 2023-12-18 at 1 29 59 PM](https://private-user-images.githubusercontent.com/9557659/291413590-4fe55a75-95aa-4330-8f6f-2ae612bee638.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5MzU0NTEsIm5iZiI6MTcwMjkzNTE1MSwicGF0aCI6Ii85NTU3NjU5LzI5MTQxMzU5MC00ZmU1NWE3NS05NWFhLTQzMzAtOGY2Zi0yYWU2MTJiZWU2MzgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIxOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMThUMjEzMjMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWM3NGJjMDA5ZDBhNDA3MzE4Njk4ZGQwNTEzNWExNTJkYTY1MTZhMjI5Mjg1YTAwOTVlOTk1MGYyMzJjNzA5YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.s3LlkWDHcZ0hM8rqtzGj-LPuPgVKgDjr_KrCwPlzgsk)

Absolutely! I will send you the demo soon! Thanks for your time!

May I know is there any community release before holiday?

efriis commented:
Yes there will be more! We usually release a few times per week

mingkang111 commented:
> Yes there will be more! We usually release a few times per week

Thank you @efriis 

I have sent the demo to [erick@langchain.dev](mailto:erick@langchain.dev). Please don't hesitate to reach out if you have any questions or need further information!

efriis commented on libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py:
```diff
@@ -0,0 +1,372 @@
+        logger.info(f"LLM API Completion:\n{completion}")
+        return completion
+
+    def send_request(
```
Is this necessary for clients to know about? Could we rename to not be public?
```suggestion
    def _send_request(
```

mingkang111 replied:
Sure. Changed to private method. Thanks!

efriis commented on docs/docs/integrations/llms/oci_model_deployment_endpoint.ipynb:
Also update to be from `langchain_community` on imports

mingkang111 replied:
Updated!

Commit:
Merge branch 'master' into feature/add_llms_integration_for_oci_datascience_model_deployment_endpoint

Commit:
fixed comments

Commit:
Fixed CI check

mingkang111 commented on libs/community/langchain_community/llms/oci_data_science_model_deployment_endpoint.py:
```diff
@@ -0,0 +1,372 @@
+    def get_lc_namespace(cls) -> List[str]:
+        """Get the namespace of the langchain object."""
+        return ["langchain", "llms", "oci_data_science_model_deployment_endpoint"]
+
```
Removed.

mingkang111 commented on libs/core/langchain_core/load/mapping.py:
Hi @efriis , Sure, I will revert this change. However, we do want to support serialization, is this mean we should create a separate PR to make the related change?  Thanks!

efriis replied:
Serialization is a separate feature here, focused on the LangSmith playground!

mingkang111 replied:
Thanks @efriis ,  I will check LangSmith for the support!

mingkang111 commented on libs/langchain/langchain/llms/__init__.py:
Reverted!

mingkang111 commented on libs/community/tests/integration_tests/llms/test_oci_model_deployment_endpoint.py:
Moved the test to unit test.

mingkang111 commented on libs/community/tests/integration_tests/llms/test_oci_model_deployment_endpoint.py:
```diff
@@ -0,0 +1,41 @@
+    )
+
+    llm = OCIModelDeploymentVLLM(endpoint=endpoint, model="my_model")
+    output = llm("This is a prompt.")
```
Fixed! Thanks for the reminder!

Commit:
reverted change in libs/langchain

Commit:
dep

Commit:
lock extra

Commit:
test bug

Commit:
resolved conflict

Commit:
lock extra

Commit:
test bugs

mingkang111 commented:
Hi @efriis ! Thanks for helping to resolve extra dependency test! I noticed some conflicts with master branch, and fixed them. Here is the screenshot of running lint check (failed before) against libs/experimental : 
![Screenshot 2023-12-19 at 9 27 51 PM](https://github.com/langchain-ai/langchain/assets/49049296/d24a3372-32e4-4171-a2e7-4ce4a849fd1e)
Reset my environment and test again:
![Screenshot 2023-12-19 at 11 58 56 PM](https://github.com/langchain-ai/langchain/assets/49049296/b62a04c1-06a6-4856-ba1c-624f3b216936)

Commit:
lock
